pkg/services/ngalert/api/api_alertmanager_guards.go

```diff
@@ -0,0 +1,157 @@
+	for name, template := range currentConfig.TemplateFiles {
+		provenance := ngmodels.ProvenanceNone
+		if prov, present := currentConfig.TemplateFileProvenances[name]; present {
+			provenance = prov
+		}
+		if provenance == ngmodels.ProvenanceNone {
+			continue // we are only interested in non none
+		}
+		found := false
+		for newName, newTemplate := range newConfig.TemplateFiles {
+			if name != newName {
+				continue
+			}
+			found = true
+			if template != newTemplate {
+				return fmt.Errorf("cannot save provisioned template '%s'", name)
+			}
+			break // we found the template and we can proceed
+		}
+		if !found {
+			return fmt.Errorf("cannot delete provisioned template '%s'", name)
+		}
+	}
+	return nil
+}
+
+func checkContactPoints(current []*apimodels.GettableApiReceiver, new []*apimodels.PostableApiReceiver) error {
+	for _, existingReceiver := range current {
+		for _, existingContactPoint := range existingReceiver.GrafanaManagedReceivers {
+			if existingContactPoint.Provenance == ngmodels.ProvenanceNone {
+				continue // we are only interested in non none
+			}
+			found, edited := false, false
+		outer:
+			for _, postedReceiver := range new {
+				for _, postedContactPoint := range postedReceiver.GrafanaManagedReceivers {
+					if existingContactPoint.UID != postedContactPoint.UID {
+						continue
+					}
+					found = true
+					if existingContactPoint.DisableResolveMessage != postedContactPoint.DisableResolveMessage {
+						edited = true
+					}
+					if existingContactPoint.Name != postedContactPoint.Name {
+						edited = true
+					}
+					if existingContactPoint.Type != postedContactPoint.Type {
+						edited = true
+					}
+					for key := range existingContactPoint.SecureFields {
+						if value, present := postedContactPoint.SecureSettings[key]; present && value != "" {
+							edited = true
+							break // it's enough to know that something was edited
+						}
+					}
+					existingSettings, err := existingContactPoint.Settings.Map()
+					if err != nil {
+						return err
+					}
+					newSettings, err := postedContactPoint.Settings.Map()
+					if err != nil {
+						return err
+					}
+					for key, val := range existingSettings {
+						if newVal, present := newSettings[key]; present {
+							if val != newVal {
+								edited = true
+								break // it's enough to know that something was edited
+							}
+						} else {
+							edited = true
+							break // it's enough to know that something was edited
+						}
+					}
+					if edited {
+						return fmt.Errorf("cannot save provisioned contact point '%s'", existingContactPoint.Name)
+					}
+					break outer
+				}
+			}
+			if !found {
+				return fmt.Errorf("cannot delete provisioned contact point '%s'", existingContactPoint.Name)
+			}
+		}
+	}
+	return nil
+}
+
+func checkMuteTimes(currentConfig apimodels.GettableUserConfig, newConfig apimodels.PostableUserConfig) error {
+	for _, muteTime := range currentConfig.AlertmanagerConfig.MuteTimeIntervals {
```

Comment:
Can you do the "two iterations" thing here too? Put all the current config ones in a map in the first step, and then do a single iteration through newConfig?

Reply:
yep, changed.